### PR TITLE
[perf] Fetch query metadata with DB calls directly

### PR DESCRIPTION
**Pull request by ranquild**

### CODE REVIEW

**Review by ranquild — COMMENTED** [review]
- dashboard link targets (cards, dashboards) in 1 call. `can_write`, `id`, `name`, `parameters`, `result_metadata`
- metrics - include the query for the source card
- load fields for native template tags that are mapped to db fields - `similar to collect-source-tables`
- do not hydrate `:segments` and `:metrics` in dashboards in `batch-fetch-query-metadatas*`

File: src/metabase/api/query_metadata.clj [hunk 1]
```diff
@@ -3,135 +3,142 @@
...
+        fk-target-tables          (api.table/batch-fetch-table-query-metadatas fk-target-table-ids)
+        tables                    (concat source-tables fk-target-tables)
+        template-tag-field-ids    (into #{} (mapcat query->template-tag-field-ids) queries)
+        query-database-ids        (into #{} (keep :database) queries)
+        database-ids              (into query-database-ids
+                                        (keep :db_id)
+                                        tables)]
+    {;; TODO: This is naive and issues multiple queries currently. That's probably okay for most dashboards,
+     ;; since they tend to query only a handful of databases at most.
+     :databases (->> (for [id database-ids]
```
Comment (ranquild) [inline src/metabase/api/query_metadata.clj:59]: @bshepherdson will it throw an exception if the database is not accessible? I think we should add a simple `get-databases` like `get-cards`.

File: src/metabase/api/query_metadata.clj [hunk 2]
```diff
@@ -1,137 +1,147 @@
...
+
+(defn- split-tables-and-legacy-card-refs [source-ids]
+  (-> (reduce (fn [m src]
+                (if-let [card-id (lib.util/legacy-string-table-id->card-id src)]
+                  (update m :cards conj! card-id)
+                  (update m :tables conj! src)))
+              {:cards  (transient #{})
+               :tables (transient #{})}
+              source-ids)
+      (update-vals persistent!)))
```
Comment (metamben) [inline src/metabase/api/query_metadata.clj:31]: Does this change actually give relevant speedup?

File: src/metabase/api/query_metadata.clj [hunk 3]
```diff
@@ -1,137 +1,147 @@
...
-        query (lib/query mp dataset-query)
-        dependents (lib/dependent-metadata query nil :question)]
-    (metadata-for-dependents (group-by :type dependents))))
+(defn batch-fetch-dashboard-metadata
+  "Fetch dependent metadata for dashboards."
+  [dashboards]
+  (let [dashcards (mapcat :dashcards dashboards)
+        cards     (for [{:keys [card series]} dashcards
+                        :let   [all (conj series card)]
+                        card all]
```
Comment (metamben) [inline src/metabase/api/query_metadata.clj:138]:
```suggestion
                        card (conj series card)]
```

**Review by metamben — APPROVED** [review]
Does this allow us to get rid of the `dependent-metadata` function or will we have to live with duplicating the logic?

**Pull request merged**